File: 57. Insert Interval.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        if not intervals:
            return [newInterval]
        n = len(intervals)
        ret = []
        i = 0
        f = False
        while i < n:
            if (not f) and newInterval.start <= intervals[i].start:
                f = True
                start, end = newInterval.start, newInterval.end
                while i < n and intervals[i].start <= end:
                    end = max(end, intervals[i].end)
                    i += 1
                if ret and start <= ret[-1].end:
                    ret[-1].end = max(ret[-1].end, end)
                else:
                    ret.append(Interval(start, end))
            else:
                ret.append(intervals[i])
                i+=1
        if not f:
            if newInterval.start <= ret[-1].end:
                ret[-1].end = max(ret[-1].end, newInterval.end)
            else:
                ret.append(newInterval)
        return ret
        

File: 57. Insert Interval.go
func insert(intervals [][]int, newInterval []int) [][]int {
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

	max := func(a, b int) int {
		if a < b {
			return b
		}
		return a
	}

	n := len(intervals)
	res := make([][]int, 0, n+1)
	for i := 0; i < n; i++ {
		if newInterval[1] < intervals[i][0] {
			res = append(res, newInterval)
			res = append(res, intervals[i:]...)
			return res
		}

		if newInterval[0] > intervals[i][1] {
			res = append(res, intervals[i])
			continue
		}

		newInterval[0] = min(newInterval[0], intervals[i][0])
		newInterval[1] = max(newInterval[1], intervals[i][1])
	}

	res = append(res, newInterval)
	return res
}
